rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp

```diff
@@ -97,26 +131,50 @@ rosbag2_storage::BagInfo SqliteStorage::info()
   return bag_info_;
 }
 
-void SqliteStorage::create_topic()
+void SqliteStorage::create_topic(const std::string & name, const std::string & type)
 {
-  initialize();
+  if (topics_.find(name) == std::end(topics_)) {
+    auto insert_topic =
+      database_->prepare_statement("INSERT INTO topics (name, type) VALUES (?, ?)");
+    insert_topic->bind(name, type);
+    insert_topic->execute_and_reset();
+    topics_.emplace(name, static_cast<int>(database_->get_last_insert_id()));
+  }
 }
 
 void SqliteStorage::prepare_for_writing()
 {
   write_statement_ = database_->prepare_statement(
-    "INSERT INTO messages (data, timestamp) VALUES (?, ?);");
+    "INSERT INTO messages (timestamp, topic_id, data) VALUES (?, ?, ?);");
 }
 
 void SqliteStorage::prepare_for_reading()
 {
-  read_statement_ =
-    database_->prepare_statement("SELECT data, timestamp FROM messages ORDER BY id;");
+  read_statement_ = database_->prepare_statement(
+    "SELECT data, timestamp, topics.name "
+    "FROM messages JOIN topics ON messages.topic_id = topics.id "
+    "ORDER BY messages.id;");
   message_result_ = read_statement_->execute_query<
-    std::shared_ptr<rcutils_char_array_t>, rcutils_time_point_value_t>();
+    std::shared_ptr<rcutils_char_array_t>, rcutils_time_point_value_t, std::string>();
   current_message_row_ = message_result_.begin();
 }
 
+void SqliteStorage::fill_topics_and_types()
+{
+  auto statement = database_->prepare_statement("SELECT name, type FROM topics ORDER BY id;");
+  auto query_results = statement->execute_query<std::string, std::string>();
+
+  for (auto result : query_results) {
```

Comment:
This could only happen if the bag is empty, i.e. contains no messages. Here the result will simply be empty which should not hurt.
I am not sure how we want to handle empty bags. Currently it would be opened, the play would initialize some things, loop over an empty list and shutdown => no harm there. So I propose to discuss this on a new issue.